### PR TITLE
Support ldap authentication between nodes

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java
Comment (cawallin) [inline presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java:27]: Can you hold off on using this until it gets merged into airlift please?

File: presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java:35]: Is there some way to not be dependent on an internal airlift version? Even just porting those changes here. Those PRs don't move very quickly, and we'll have to continuously rebase them whenever airlift is released again.
(btw, https://github.com/airlift/airlift/pull/473 has a failing build in travis)

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java
Comment (cawallin) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java:45]: Isn't both user and userAgent redundant?

File: presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java
Comment (cawallin) [inline presto-cli/src/main/java/com/facebook/presto/cli/ClientOptions.java:87]: David explicitly said he didn't want to have this, I don't think we should add it.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java:216]: Irritating as it is to modify tempto, I think we should do that (either to echo the password or export an environment variable when running the CLI) so that we don't go directly against David's suggestion.

**Review by cawallin — CHANGES_REQUESTED** [review]

File: presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java
Comment (arhimondr) [reply presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java:27]: Then i should copy this class to 3 modules. `presto-product-tests`, `presto-jdbc`, `presto-main`. Furthermore, i still need airlift patched in order to register this filter to all of the internal http clients in `presto-main.`.

File: presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java:35]: This failure is intermittent. Unfortunately there is no other way to register the basic authentication filter to all of the internal HTTP clients. I know what the consequences this introduce. And if there was a way to do that without modifying Airlift - i would definitely do so.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java
Comment (arhimondr) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java:45]: `userAgent` - is the name of the client app the Presto server is being queries (e.g. `presto-cli`, `presto-jdbc`, etc.). `user` - is the login user.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java
Comment (arhimondr) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java:216]: Implementing it with echo is not convenient. We need to hack tempto, we need to hack `run_on_docker` script. This PR has a lot of dependencies. It depends on ldap authentication, on ssl in opensource jdbc, on ssl internal communication. So, it is long way until it will land on the master. Hacking around the stuffs is painful, because you will need to rebase all that changes all the way.

I believe that having hidden parameter for password is fine. All the cli clients (mysql, mssql, hive, oracle, postgree) have so. We can try to convince David to merge it. Because it is really convenient for testing. If he insisted  - then we would hack around. I just don't want to run ahead of time.

File: presto-product-tests/conf/docker/common/compose-commons.sh
Comment (maciejgrzybek) [inline presto-product-tests/conf/docker/common/compose-commons.sh]: `export TEMPTO_CONFIG_YAML=${TEMPTO_CONFIG_YAML:""}`

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java
Comment (maciejgrzybek) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java:136]: Didn't you forget about prompt here?
Hidden `-P` but what about `--password` which existed before?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: `requireNonNull` is not needed.

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java
Comment (arhimondr) [reply presto-cli/src/main/java/com/facebook/presto/cli/Console.java:136]: `getPassword()` does both. If the `-P` is specified - it takes the password from parameter. If the `--password` flag is specified - it shows the prompt.

**Comment by arhimondr** [comment]
@maciejgrzybek @cawallin Any other comments guys? =)

File: presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java:35]: What do you think about adding those patches to the patches bookkeeping? Maybe on a separate tab? Can you make sure to monitor airlift releases as you rebase this PR? I think we should get these questions settled before this is merged.

File: presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java
Comment (cawallin) [reply presto-cli/src/main/java/com/facebook/presto/cli/LdapRequestFilter.java:27]: :/ ok. I see how it's not possible to do this without airlift changes.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java
Comment (cawallin) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java:216]: OK, sounds fine.

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java
Comment (cawallin) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: I think this requires the user to always be specified, you should change this checkstate if that's the case

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java
Comment (cawallin) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/HttpClientCreator.java:45]: What do you think about renaming it to databaseUser?

**Review by cawallin — APPROVED** [review]
lgtm % tiny comments -- the checkState is the more critical one, I don't insist too strongly about the databaseUser one

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java
Comment (arhimondr) [reply presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: I'm just going to remove this check. `clientOptions.user` just couldn't be null. It has a default.

**Comment by arhimondr** [comment]
> What do you think about renaming it to databaseUser?

The property is called `USER`. So, i think it is better to call it `user` to keep the naming consistent.

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java
Comment (maciejgrzybek) [reply presto-cli/src/main/java/com/facebook/presto/cli/Console.java:136]: What if we specify both, -P and --password?

Comment (arhimondr) [reply presto-cli/src/main/java/com/facebook/presto/cli/Console.java:136]: Than the prompt will be shown

**Pull request closed without merging**
